---
title: "Gnocchi analysis"
format: html
---

```{r warning = FALSE, message = FALSE}
library(readr)
library(janitor)
library(dplyr)
library(patchwork)
library(tidyr)
library(purrr)
library(stringr)
library(ggplot2)
library(readxl)
library(conflicted)
library(seqminer)

devtools::load_all("~/coloc")

conflicts_prefer(dplyr::filter)
conflicts_prefer(dplyr::select)
```

```{r}
eqtl_data_file <- "data/QTD000110.cc.tsv.gz"
pqtl_data_file <- "data/QTD000584.cc.tsv.gz"
eqtl_metadata_file <- "data/metadata/gene_counts_Ensembl_105_phenotype_metadata.tsv.gz"
pqtl_metadata_file <- "data/metadata/SomaLogic_Ensembl_96_phenotype_metadata.tsv.gz"

eqtl_metadata <- read_tsv(eqtl_metadata_file, show_col_types = FALSE)

eqtl_metadata |>
  dplyr::filter(gene_name == "IL18R1") |>
  mutate(tss = if_else(strand == 1, gene_start, gene_end)) |>
  rowwise() |>
  mutate(region = paste0(chromosome, ":", tss - 1e5, "-", tss + 1e5)) |>
  dplyr::select(chromosome, gene_id, region, tss)

eqtl_col_names <- colnames(read_tsv("data/eqtl-catalogue/sumstats/QTD000373.cc.tsv.gz", n_max = 1, 
                           show_col_types = FALSE))

eqtl_data <- tabix.read.table("data/eqtl-catalogue/sumstats/QTD000373.cc.tsv.gz", 
                              paste0(2, ":1-2147483647")) |>
  as_tibble() |>
  setNames(eqtl_col_names) |>
  filter(molecular_trait_id == "ENSG00000115604") |>
  filter(chromosome == 2) |>
  filter(
    position >= 102211529 &
    position <= 102411529
  ) |>
  select(-rsid) |>
  distinct() |>
  mutate(id = paste(chromosome, position, sep = ":")) |>
  group_by(id) |>
  mutate(row_count = n()) |>
  ungroup() |>
  filter(row_count == 1) |>
  filter(!is.nan(se)) |>
  filter(!is.na(se)) |>
  select(molecular_trait_id, variant, maf, beta, se, an, position) |>
  mutate(
    maf = as.numeric(maf),
    beta = as.numeric(beta),
    se = as.numeric(se),
    an = as.numeric(an)
  )

pqtl_data <- tabix.read.table("data/QTD000584.cc.tsv.gz", paste0(2, ":1-2147483647")) |>
  as_tibble() |>
  # All the column names are the same.
  setNames(eqtl_col_names) |>
  filter(molecular_trait_id == "IL18R1.14079.14.3..1") |>
  filter(chromosome == 2) |>
  filter(
    position >= 102211529 &
    position <= 102411529
  ) |>
  select(-rsid) |>
  distinct() |>
  mutate(id = paste(chromosome, position, sep = ":")) |>
  group_by(id) |>
  mutate(row_count = n()) |>
  ungroup() |>
  filter(row_count == 1) |>
  filter(!is.nan(se)) |>
  filter(!is.na(se)) |>
  select(molecular_trait_id, variant, maf, beta, se, an, position) |>
  mutate(
    maf = as.numeric(maf),
    beta = as.numeric(beta),
    se = as.numeric(se),
    an = as.numeric(an)
  )

eqtl_data <- eqtl_data |>
  filter(variant %in% pqtl_data$variant)
pqtl_data <- pqtl_data |>
  filter(variant %in% eqtl_data$variant)

eqtl_dataset <- list(
  varbeta = eqtl_data$se^2,
  N = eqtl_data$an * 2,
  MAF = eqtl_data$maf ,
  type = "quant",
  beta = eqtl_data$beta,
  snp = eqtl_data$variant,
  position = eqtl_data$position
)

pqtl_dataset <- list(
  varbeta = pqtl_data$se^2,
  N = pqtl_data$an * 2,
  MAF = pqtl_data$maf ,
  type = "quant",
  beta = pqtl_data$beta,
  snp = pqtl_data$variant,
  position = pqtl_data$position
)

devtools::load_all("~/coloc")

gnocchi_data <- readr::read_tsv("/home/jp2045/coloc-estimated-eqtl-priors/data/gnocchi-windows.bed", 
                                  col_names = FALSE, show_col_types = FALSE)
colnames(gnocchi_data) <- c("chromosome", "start_pos", "end_pos", "score")

coloc::plot_dataset(eqtl_dataset)
abline(v = 102311529)
coloc::plot_dataset(pqtl_dataset)

suppressWarnings(coloc::coloc.abf(eqtl_dataset, pqtl_dataset)$summary)
suppressWarnings(coloc::coloc.abf(eqtl_dataset, pqtl_dataset, tss1 = 102311529)$summary)

suppressWarnings(coloc.abf(eqtl_dataset, pqtl_dataset, score_data1 = gnocchi_data, chrom = 2)$summary)
```
