---
title: "GWAS-eQTL colocalisation analysis"
format: html
editor: source
---

```{r warning = FALSE, message = FALSE}
renv::load(here::here())

library(readr)
library(dplyr)
library(tidyr)
library(ggplot2)
library(forcats)
library(yaml)
library(glue)
library(extraDistr)

source(here::here("code/coloc-utils.R"))

devtools::load_all("~/coloc")
```

```{r load-data}
config <- read_yaml(here::here("config.yaml"))
gwas_ids_eqtl_ids <- config$gwas_eqtl_coloc_ids
gwas_eqtl_colocs <- expand_grid(
  chr = 1:22,
  gwas_id_eqtl_id = gwas_ids_eqtl_ids
  ) |>
  separate_wider_delim(gwas_id_eqtl_id, "-", names = c("gwas_id", "eqtl_id")) |>
  rowwise() |>
  mutate(file = here::here(
    glue("output/data/gwas-eqtl-coloc-{gwas_id}-{eqtl_id}-{chr}.rds")
  )) |>
  mutate(data = list(read_rds(file))) |>
  unnest(data)
```

```{r}
gwas_eqtl_colocs |>
  filter(gwas_id == "AUTOIMMUNE") |>
  filter(PP.H4.abf_unif > 0.8) |>
  select(gwas_id, eqtl_id, gene_name, PP.H4.abf_unif) |>
  print(n = 100)
```

```{r}
gwas_eqtl_colocs |>
  filter(gwas_id == "AUTOIMMUNE") |>
  filter(PP.H4.abf_unif > 0.8) |>
  count(gene_name) |>
  arrange(desc(n)) |>
  print(n = 47)
```

```{r}
gwas_eqtl_colocs |>
  filter(gwas_id == "AUTOIMMUNE") |>
  filter(PP.H4.abf_unif > 0.8) |>
  count(eqtl_id) |>
  arrange(desc(n)) |>
  print(n = 47)
```

```{r}
gwas_eqtl_colocs |>
  filter(gwas_id == "T2D_WIDE") |>
  filter(PP.H4.abf_unif > 0.8) |>
  select(gwas_id, eqtl_id, gene_name, PP.H4.abf_unif) |>
  print(n = 50)
```

```{r}
gwas_eqtl_colocs |>
  select(gene_name, gwas_id, starts_with("PP.H4.abf")) |>
  pivot_longer(
    -c(gene_name, gwas_id), 
    names_to = "prior",
    values_to = "pp_h4"
  ) |>
  mutate(sig_coloc = pp_h4 > 0.8) |>
  summarise(n_sig = sum(sig_coloc), .by = c(prior, gwas_id)) |>
  mutate(is_unif = prior == "PP.H4.abf_unif") |>
  mutate(prior = fct_reorder(factor(prior), n_sig, mean)) |>
  ggplot(aes(prior, n_sig, fill = is_unif)) + 
  geom_col() + 
  coord_flip() + 
  labs(
    y = "Number of signifcant (Pr(H4) > 0.8) colocalisations", 
    x = "Prior type"
  ) +
  facet_wrap(~gwas_id, scales = "free_x") + 
  theme_bw() + 
  theme(legend.position = "none")
```

CRIP3 colocalises in monocytes
CHP1 colocalises in eQTLGen
PIGV colocalises in monocytes

```{r}
gwas_eqtl_colocs |>
  mutate(
    sig_abc_score_gwas = PP.H4.abf_abc_score_gwas > 0.8,
    sig_unif = PP.H4.abf_unif > 0.8
  ) |>
  arrange(gwas_id) |>
  filter(sig_abc_score_gwas != sig_unif) |>
  select(gwas_id, eqtl_id, gene_name,
         PP.H4.abf_unif, PP.H4.abf_abc_score_gwas, 
         prop_sig_pp_h4_2)
```

```{r}
gwas_eqtl_colocs |>
  mutate(
    sig_abc_score_gwas = PP.H4.abf_abc_score_gwas > 0.8,
    sig_unif = PP.H4.abf_unif > 0.8
  ) |>
  rowwise() |>
  mutate(
    q25 = q_pp_h4_2[[2]],
    q75 = q_pp_h4_2[[6]]
  ) |>
  ungroup() |>
  arrange(gwas_id) |>
  filter(sig_abc_score_gwas != sig_unif) |>
  select(eqtl_id, gene_name,
         PP.H4.abf_unif, PP.H4.abf_abc_score_gwas, 
         q25, q75)
```

```{r}
gwas_eqtl_colocs |>
  mutate(
    sig_gwas_tss_eqtlgen = PP.H4.abf_gwas_tss_eqtlgen > 0.8,
    sig_unif = PP.H4.abf_unif > 0.8
  ) |>
  arrange(gwas_id) |>
  filter(sig_gwas_tss_eqtlgen != sig_unif) |>
  select(gwas_id, eqtl_id, gene_name,
         PP.H4.abf_unif, PP.H4.abf_gwas_tss_eqtlgen, 
         prop_sig_pp_h4_2) |>
  print(n = 80)
```

```{r}
gwas_eqtl_colocs |>
  mutate(
    sig_gwas_tss_eqtlgen = PP.H4.abf_eqtl_tss_eqtlgen > 0.8,
    sig_unif = PP.H4.abf_unif > 0.8
  ) |>
  arrange(gwas_id) |>
  filter(sig_gwas_tss_eqtlgen != sig_unif) |>
  select(gwas_id, eqtl_id, gene_name,
         PP.H4.abf_unif, PP.H4.abf_eqtl_tss_eqtlgen, 
         prop_sig_pp_h4_2) |>
  print(n = 80)
```

```{r}
gwas_eqtl_colocs |>
  mutate(
    sig_gwas_tss_eqtlgen = PP.H4.abf_abc_score_eqtl > 0.8,
    sig_unif = PP.H4.abf_unif > 0.8
  ) |>
  arrange(gwas_id) |>
  filter(sig_gwas_tss_eqtlgen != sig_unif) |>
  select(gwas_id, eqtl_id, gene_name,
         PP.H4.abf_unif, PP.H4.abf_abc_score_eqtl, 
         prop_sig_pp_h4_2) |>
  print(n = 80)
```

```{r}
gwas_eqtl_colocs |>
  mutate(
    sig_gwas_rand = PP.H4.abf_rand_gwas > 0.8,
    sig_unif = PP.H4.abf_unif > 0.8
  ) |>
  arrange(gwas_id) |>
  filter(sig_gwas_rand != sig_unif) |>
  select(gwas_id, eqtl_id, gene_name,
         PP.H4.abf_unif, PP.H4.abf_rand_gwas)
```

```{r}
gwas_eqtl_colocs |>
  mutate(
    sig_gwas_rand = PP.H4.abf_rand_eqtl > 0.8,
    sig_unif = PP.H4.abf_unif > 0.8
  ) |>
  arrange(gwas_id) |>
  filter(sig_gwas_rand != sig_unif) |>
  select(gwas_id, eqtl_id, gene_name,
         PP.H4.abf_unif, PP.H4.abf_rand_eqtl)
```

```{r}
gwas_eqtl_colocs |>
  filter(gene_name == "GINS2") |>
  select(gwas_id, eqtl_id, gene_name, 
        PP.H4.abf_unif, PP.H4.abf_abc_score_gwas)
```

```{r}
gwas_eqtl_colocs |>
  arrange(desc(var_pp_h4_2)) |>
  select(gene_name, var_pp_h4_2) |>
  print(n = 20)
```

```{r}
gwas_eqtl_colocs |>
  ggplot(aes(PP.H4.abf_unif, prop_sig_pp_h4_2)) + 
  geom_point() + 
  geom_vline(xintercept = 0.8, col = "red") + 
  geom_hline(yintercept = 0.8, col = "blue") + 
  theme_bw()
```

```{r}
gwas_eqtl_colocs |>
  ggplot(aes(PP.H4.abf_unif, prop_sig_pp_h4_1)) + 
  geom_point() + 
  geom_vline(xintercept = 0.8, col = "red") + 
  theme_bw()
```

```{r}
gwas_eqtl_colocs |>
  ggplot(aes(PP.H4.abf_unif, PP.H4.abf_abc_score_gwas, col = var_pp_h4_2)) + 
  geom_point() + 
  geom_hline(yintercept = 0.8, col = "red") + 
  geom_vline(xintercept = 0.8, col = "red") + 
  theme_bw()
```

```{r}
gwas_eqtl_colocs |>
  ggplot(aes(PP.H4.abf_unif, PP.H4.abf_abc_score_gwas)) + 
  geom_point() + 
  geom_hline(yintercept = 0.8, col = "red") + 
  geom_vline(xintercept = 0.8, col = "red") + 
  theme_bw()
```

```{r}
gwas_eqtl_colocs |>
  ggplot(aes(PP.H4.abf_unif, PP.H4.abf_gwas_tss_eqtlgen)) + 
  geom_point() + 
  geom_abline(linetype = "dotted") + 
  geom_hline(yintercept = 0.8, col = "red") + 
  geom_vline(xintercept = 0.8, col = "red") + 
  theme_bw()
```

```{r}
gwas_eqtl_colocs |>
  ggplot(aes(PP.H4.abf_unif, PP.H4.abf_eqtl_tss_eqtlgen)) + 
  geom_point() + 
  geom_abline(linetype = "dotted") + 
  geom_hline(yintercept = 0.8, col = "red") + 
  geom_vline(xintercept = 0.8, col = "red") + 
  theme_bw()
```


```{r}
gwas_eqtl_colocs |>
  ggplot(aes(PP.H4.abf_unif, PP.H4.abf_rand_gwas)) + 
  geom_point() + 
  geom_abline(linetype = "dotted") + 
  geom_hline(yintercept = 0.8, col = "red") + 
  geom_vline(xintercept = 0.8, col = "red") + 
  theme_bw()
```

```{r}
#| eval: false

# "Region: 11:207118-1207118"
# "Gene ID: ENSG00000177030"
# "Gene name: DEAF1"

eqtl_metadata <- read_tsv(
  "data/metadata/gene_counts_Ensembl_105_phenotype_metadata.tsv.gz",
  show_col_types = FALSE
)
manifest_data <- read_tsv(
  "data/finngen/finngen-manifest.tsv",
  show_col_types = FALSE
)

#I9_HYPTENS QTD000549 ZNF100  
width <- 5e5
coloc_metadata <- eqtl_metadata |>
  filter(gene_type == "protein_coding") |>
  mutate(tss = if_else(strand == 1, gene_start, gene_end)) |>
  rowwise() |>
  mutate(
    start_pos = max(tss - width, 1),
    end_pos = tss + width
  ) |>
  mutate(region = paste0(chromosome, ":", start_pos, "-", end_pos)) |>
  select(gene_name, region, gene_id, chromosome, start_pos, end_pos, tss, gene_name) |>
  ungroup() |>
  filter(!is.na(gene_id))

coloc_metadata |>
  filter(gene_name == "ZNF100") |>
  select(gene_id, chromosome, start_pos, end_pos, tss)

gwas_id <- "I9_HYPTENS"
range <- "19:21267579-22267579"
gwas_data <- seqminer::tabix.read.table("data/finngen/I9_HYPTENS.gz", range) |>
  rename(
    rsid = rsids,
    chromosome = chrom,
    position = pos,
    beta = beta,
    se = sebeta,
    maf = af_alt
  ) |>
  mutate(
    molecular_trait_id = gwas_id,
    variant = paste0("chr", chromosome, "_", position, "_", ref, "_", alt)
  ) |>
  select(-c(af_alt_cases, af_alt_controls, ref, alt)) |>
  left_join(
    manifest_data |>
      select(phenocode, num_cases, num_controls),
    by = join_by(molecular_trait_id == phenocode)
  ) |>
  mutate(N = num_cases + num_controls) |>
  select(-c(num_cases, num_controls)) |>
  prepare_coloc_dataset()

eqtl_data <- seqminer::tabix.read.table("data/eqtl-catalogue/sumstats/QTD000549.cc.tsv.gz", range) |> 
  as_tibble() |>
  setNames(eqtl_catalouge_colnames) |>
  filter(molecular_trait_id == "ENSG00000197020") |>
  prepare_coloc_dataset()

eqtl_data <- eqtl_data |>
  filter(variant %in% gwas_data$variant)
gwas_data <- gwas_data |>
  filter(variant %in% eqtl_data$variant)

eqtl_dataset <- list(
  varbeta = eqtl_data$se^2,
  N = eqtl_data$an / 2,
  MAF = eqtl_data$maf,
  type = "quant",
  beta = eqtl_data$beta,
  snp = eqtl_data$variant,
  position = eqtl_data$position
)

gwas_dataset <- list(
  varbeta = gwas_data$se^2,
  N = gwas_data$N,
  MAF = gwas_data$maf,
  type = "cc",
  beta = gwas_data$beta,
  snp = gwas_data$variant,
  position = gwas_data$position
)

abc_score_data <- read_tsv("data/abc-score-data.txt.gz", show_col_types = FALSE)
abc_score_prior_weights <- compute_abc_prior_weights(
  eqtl_dataset$position, 19, "ZNF100", abc_score_data
)
coloc.abf(
  eqtl_dataset, 
  gwas_dataset,
  prior_weights2 = abc_score_prior_weights
)
eqtlgen_density_data <- read_rds("output/densities/eqtlgen.rds")
eqtl_prior_weights_eqtlgen <- compute_eqtl_tss_dist_prior_weights(
  eqtl_dataset$position, 77062558, eqtlgen_density_data
)
plot(eqtl_dataset$position, abc_score_prior_weights)

coloc::plot_dataset(eqtl_dataset)
coloc::plot_dataset(gwas_dataset)

pp_h4 <- numeric(500)
n <- length(eqtl_dataset$beta)
for (i in 1:500) {
  p <- as.vector(rdirichlet(1, rep(1, n)))
  pp_h4[[i]] <- coloc.abf(
    eqtl_dataset, 
    gwas_dataset, 
    prior_weights1 = p,
    prior_weights2 = p
  )$summary["PP.H4.abf"]
}
hist(pp_h4)
var(pp_h4)
IQR(pp_h4)
```